runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -471,61 +472,84 @@ private ILogData cacheLoadAndGet(@NonNull Cache<Long, ILogData> cache, long addr
-                                // can't be used here because it throws a UnrecoverableCorfuInterruptedError
-                                try {
-                                    return future.join();
-                                } catch (CompletionException ex) {
-                                    Throwable cause = ex.getCause();
-                                    if (cause instanceof RuntimeException) {
-                                        throw (RuntimeException) cause;
-                                    } else {
-                                        throw new RuntimeException(cause);
-                                    }
-                                }
-                            })
-                            .max(Comparator.naturalOrder()).get();
-                    return new Token(e.getLayout().getEpoch(), trimMark);
-                });
+        return getTrimMark(true);
+    }
+
+    /**
+     * Get the first address in the address space.
+     *
+     * @param retry whether to do retry on certain failures
+     * @return a token with epoch and first address
+     */
+    public Token getTrimMark(boolean retry) {
+        return layoutHelper(Utils::getTrimMark, !retry);
     }
 
     /**
      * Get the log's tail, i.e., last address in the address space.
      */
     public Long getLogTail() {
-        return layoutHelper(
-                e -> Utils.getLogTail(e.getLayout(), runtime));
+        return layoutHelper(Utils::getLogTail);
     }
 
     /**
      * Get all tails, includes: log tail and stream tails.
      */
     public TailsResponse getAllTails() {
-        return layoutHelper(
-                e -> Utils.getAllTails(e.getLayout(), runtime));
+        return layoutHelper(Utils::getAllTails);
     }
 
     /**
      * Get log address space, which includes:
      * 1. Addresses belonging to each stream.
      * 2. Log Tail.
      *
-     * @return
+     * @return log address space
      */
     public StreamsAddressResponse getLogAddressSpace() {
-        return layoutHelper(
-                e -> Utils.getLogAddressSpace(e.getLayout(), runtime));
+        return layoutHelper(Utils::getLogAddressSpace);
+    }
+
+    /**
+     * Get the maximum committed log tail from all log units.
+     *
+     * @return the maximum committed log tail
+     */
+    public long getCommittedTail() {
+        return layoutHelper(Utils::getCommittedTail, true);
+    }
+
+    /**
+     * Commit the addresses in the range by first inspecting the addresses
+     * and if data does not exist in log, hole fill the address. This is
+     * used by management agent for log consolidation.
+     *
+     * @param start start of address range, inclusive
+     * @param end   end of address range, inclusive
+     */
+    public void commit(long start, long end) {
+        if (start >= end) {
+            return;
+        }
+
+        ContiguousSet<Long> range = ContiguousSet.create(
```

Comment:
Are there any advantages of using this rather than a simple ArrayList/ImmutableList, for example?

Reply:
Here we need a fast way to create a contiguous list/set given by two endpoints.